### PR TITLE
Unify function output

### LINKED ISSUE #421: pspm_write_channel variable assignment
Recent development of PsPM aims at generalising the variable `options` in all pspm functions. `channel_action`, one of the subfields of `options`, has been generalised for every function except for `pspm_write_channel`. For every other function, `channel_action` is a subfield in `options`, and it can be whether mandatory or optional. For `pspm_write_channel`, `channel_action` is an individual variable which is parallel to `options`. This is an exception for `channel_action` and we may want to reconsider where to put it.

We have discussed about this issue in this week's group meeting. Our colleagues have provided their suggestions:
* One of our colleagues suggested `channel_action` should be prioritised in `pspm_write_channel` because it plays a vital role in the process of `pspm_write_channel`. As an individual variable, it will cause attention of users and force them to always provide value for it.
* Another colleague suggested `channel_action` could be renamed in `pspm_write_channel`. This is because throughout PsPM, `options` always tends to have the field `channel_action` when applicable. Users may become confused whether to specify such value in the individual variable of `pspm_write_channel` or the field of `options` in `pspm_write_channel`.

### LINKED ISSUE #519: Streamline function output
### Feature Description
- make a decision on static vs dynamic output
Suggestion for static output: 
- sts should be provided for all internally called functions that produce output, to offer a systematic way to handle problems without interrupting batch processing
- functions that are called like scripts for their side effect don't need sts

### LINKED ISSUE #624: Improve pspm_write_channel
### Feature Description

Currently, pspm_write_channel, overwrites the last channel of same type if channel_action is 'replace'. However, when channels are defined by numerical index in the calling function, it would be more appropriate to overwrite the input channel.

## Steps to solve this
1. Identify all functions that use `pspm_write_channel`, allow `channel_action`to be `replace`, and do not create channels of new type. 
2. Check whether these functions pass correct channel numbers to `pspm_write_channels`. Channel numbers should correspond to those that were retrieved from the file with a `pspm_load_channel` function call, or otherwise within the calling function.
3. Compile a list of functions that fulfill 1 and not 2.
4. Implement correct assignment of channel numbers in these functions.

### PR DESCRIPTION
**Pull request by dominikbach**
Fixes #624. Fixes #519. Fixes #421. 

## Changes proposed in this pull request:
1. Functions that write data into channels always return `[sts, channel_index]` and potentially further output arguments.
2. Functions that create new files always return `[sts, newfilename]` and potentially further output arguments.
3. Functions always work on one data file only unless explicitly designed to work on multiple files (such as `pspm_exp`).
4. `pspm_write_channel` is always called with an index of the channels that were used to create the new channel data. `pspm_write_channel` will then check whether these channels will be overwritten (if action is `replace` and the channels have the same type and units) or whether new channels are created.
5. `varargout` was removed wherever possible. Some widely used helper functions remain (e.g. `pspm_overwrite` and `pspm_prepdata`) and will be addressed in new pull requests.
6. GUIs and tests were changed accordingly. 
7. Some inconsistencies remain in `cfg_run` functions and will be addressed in new PR.

### Function output was changed:
```
pspm_combine_markerchannels
pos_of_channel
pspm_convert_au2unit
pspm_convert_ecg2hb
pspm_convert_hb2hp
pspm_convert_ppg2hb
pspm_emg_pp
pspm_find_sounds
pspm_resp_pp
pspm_scr_pp
pspm_pp
pspm_ren
pspm_remove_epochs
pspm_import
pspm_trim
pspm_glm
pspm_interpolate
```
### Write channel behaviour changed
```
pspm_convert_au2unit
pspm_convert_gaze
pspm_pupil_correct_eyelink
pspm_pupil_pp
pspm_scr_pp
pspm_remove_epochs
```
### Option to work on multiple files removed
```
pspm_scr_pp
pspm_merge
pspm_trim
pspm_import
```
### Varargout removed
```
pspm_merge
pspm_split_sessions
pspm_ren
pspm_trim
pspm_dcm
pspm_glm_recon
pspm_sf
pspm_sf_auc
pspm_sf_dcm
pspm_sf_mp
pspm_sf_scl
pspm_transfer_function
pspm_version
pspm_write_channel
```
### Function output unchanged
```
pspm_convert_ecg2hb_amri
pspm_pupil_correct_eyelink
pspm_pupil_pp
```
### Bugs fixed
```
pspm_load_channel
```
### Other changes
- `pspm_ren` was renamed to `pspm_rename` for enhanced transparency
- `pspm_rename` now uses `pspm_load_data` to save data and checks whether new file already exists, via `options.overwrite`
- improved `pspm_write_channel`, now uses `pspm_select_channels` for checks
- new function `pspm_check_data`, used in `pspm_load_data` and `pspm_write_channel`
- improved `pspm_select_channels`

### CODE REVIEW

**Comment by teddychao** [comment]
Hi Dom,

Thank you for the update.

My flight is tomorrow and I am preparing for several stuffs. Is it ok if I review this next week?

Best wishes
Teddy

From: Dominik Bach ***@***.***>
Date: Friday, 19 April 2024 at 09:29
To: bachlab/PsPM ***@***.***>
Cc: Teddy ***@***.***>, Review requested ***@***.***>
Subject: Re: [bachlab/PsPM] Unify function output (PR #689)

@dominikbach<https://github.com/dominikbach> requested your review on: #689<https://github.com/bachlab/PsPM/pull/689> Unify function output.

**Review by teddychao — CHANGES_REQUESTED** [review]
I have reviewed some functions.. I try to finish them in these two days.

File: src/pspm_cfg/pspm_cfg_run_pp_heart_data.m [hunk 1]
```diff
@@ -23,20 +23,20 @@
                                      {'semi', 'twthresh'});
         options = pspm_update_struct(options, job, 'channel_action');
         % call function
-        [sts, winfo] = pspm_convert_ecg2hb(fn, options);
+        [sts, outchannel] = pspm_convert_ecg2hb(fn, options);
       case 'ecg2hb_amri'
         options = pp_field.opt;
         options.channel = chan;
         options = pspm_update_struct(options, job, 'channel_action');
         winfo = struct();
```
Comment (teddychao) [inline src/pspm_cfg/pspm_cfg_run_pp_heart_data.m:31]: Think this line can be removed.

File: src/pspm_find_sounds.m [hunk 2]
```diff
@@ -1,12 +1,12 @@
-function [sts, infos] = pspm_find_sounds(varargin)
+function [sts, outchannel, outinfos] = pspm_find_sounds(fn, options)
 % ● Description
 %   pspm_find_sounds finds and if required analyzes sound events in a pspm file.
 %   A sound is accepted as event if it is longer than 10 ms and events are
 %   recognized as different if they are at least 50 ms appart.
 % ● Format
-%   [sts, infos] = pspm_find_sounds(file,options)
+%   [sts, channel_index, info] = pspm_find_sounds(fn, options)
```
Comment (teddychao) [inline src/pspm_find_sounds.m:7]: Maybe consistent variable naming is required here too

File: src/pspm_find_sounds.m [hunk 3]
```diff
@@ -20,19 +20,17 @@
 % │                 each channel.  In this case, set the option 
 % │                 'channel_action' to 'add',  to store each
 % │                 resulting channel separately.
-% ├.channel_action: ['none'/'add'/'replace'] if not set to 'none'
+% ├.channel_action: ['add'/'replace']
```
Comment (teddychao) [inline src/pspm_find_sounds.m:23]: We may want to discuss about whether we would like to keep the optional value `none` for `channel_action` ?

**Review by teddychao — CHANGES_REQUESTED** [review]

File: src/pspm_check_data.m
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_check_data.m:1]: 1. the output variable `data` is an altered version of the input with legacy and format  conversions (e.g. line 53) so should be kept.
2. Numbering in comments removed now.

File: src/pspm_cfg/pspm_cfg_rename.m [hunk 4]
```diff
@@ -8,14 +8,14 @@
...
-%filename.filter  = '\.mat$';
+filename.filter  = '\.mat$';
 filename.help    = {'Choose name of original file.'};
 
 newfilename         = cfg_entry;
 newfilename.name    = 'New File Name';
 newfilename.tag     = 'newfilename';
 newfilename.strtype = 's';
-%newfilename.num     = [1 1];
+newfilename.num     = [1 1];
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_cfg/pspm_cfg_rename.m:18]: fixed now.

File: src/pspm_cfg/pspm_cfg_run_find_sounds.m
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_cfg/pspm_cfg_run_find_sounds.m:1]: We always want to create a channel (i.e. never give the GUI user the option not to). But this was incorrectly implemented in the GUI. It should be fixed now.

File: src/pspm_sf_mp.m [hunk 5]
```diff
@@ -1,10 +1,10 @@
-function varargout = pspm_sf_mp(model, options)
+function [sts, out] = pspm_sf_mp(model, options)
 % ● Description
 %   pspm_sf_mp does the inversion of a DCM for SF of the skin conductance, using
 %   a matching pursuit algorithm, and f_SF for the forward model
 %   the input data is assumed to be in mcS, and sampling rate in Hz
 % ● Format
-%   function out = pspm_sf_mp(scr, sr, options)
+%   [sts, mp] = pspm_sf_mp(model, options)
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_sf_mp.m:7]: This is just stylistic and does not impair functionality.

File: src/pspm_pupil_correct_eyelink.m [hunk 6]
```diff
@@ -15,7 +15,7 @@
 %   'channel_action', it will either replace an existing preprocessed pupil
 %   channel or add it as new channel to the provided file.
 % ● Format
-%   [sts, out_channel] = pspm_pupil_correct_eyelink(fn, options)
+%   [sts, channel_index] = pspm_pupil_correct_eyelink(fn, options)
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_pupil_correct_eyelink.m:18]: Just stylistic.

File: src/pspm_convert_ecg2hb_amri.m [hunk 7]
```diff
@@ -9,8 +9,8 @@
 %   it will either replace an existing heartbeat channel or add it as a new
 %   channel to the provided file.
 % ● Format
-%   [sts, out_channel] = pspm_convert_ecg2hb_amri(fn)
-%   [sts, out_channel] = pspm_convert_ecg2hb_amri(fn, options)
+%   [sts, channel_index] = pspm_convert_ecg2hb_amri(fn)
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_convert_ecg2hb_amri.m:12]: Just stylistic.

File: src/pspm_convert_hb2hp.m [hunk 8]
```diff
@@ -1,9 +1,9 @@
-function [sts, infos] = pspm_convert_hb2hp(fn, sr, options)
+function [sts, outchannel] = pspm_convert_hb2hp(fn, sr, options)
 % ● Description
 %   pspm_convert_hb2hp transforms heart beat data into an interpolated heart
 %   rate signal and adds this as an additional channel to the data file
 % ● Format
-%   sts = pspm_convert_hb2hp(fn, sr, options)
+%   [sts, channel_index] = pspm_convert_hb2hp(fn, sr, options)
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_convert_hb2hp.m:6]: Just stylistic.

File: src/pspm_convert_ppg2hb.m [hunk 9]
```diff
@@ -7,7 +7,7 @@
 %   identified as heartbeat maximas and a heartbeat channel is then
 %   generated from these.
 % ● Format
-%   [ sts, outinfo ] = pspm_convert_ppg2hb( fn, options )
+%   [sts, channel_index] = pspm_convert_ppg2hb( fn, options )
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_convert_ppg2hb.m:10]: Just stylistic.

File: src/pspm_sf_dcm.m [hunk 10]
```diff
@@ -1,10 +1,10 @@
-function varargout = pspm_sf_dcm(model, options)
+function [sts, out] = pspm_sf_dcm(model, options)
 % ● Description
 %   pspm_sf_dcm does dynamic causal modelling for SF of the skin conductance
 %   uses f_SF and g_Id
 %   the input data is assumed to be in mcS, and sampling rate in Hz
 % ● Format
-%   function out = pspm_sf_dcm(model, options)
+%   [sts, dcm] = pspm_sf_dcm(model, options)
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_sf_dcm.m:7]: This is a stylistic issue, no change.

File: src/pspm_sf_auc.m
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_sf_auc.m:1]: This is a stylistic issue, no change.

File: src/pspm_version.m
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_version.m:1]: removed now.

File: src/pspm_merge.m [hunk 11]
```diff
@@ -26,106 +24,73 @@
...
-  infos = cell(2,1);
-  data = cell(2,1);
-  for iNum = 1:2
-    [sts_load_data, infos{iNum}, data{iNum}] = pspm_load_data(infile{iNum}{iFile});
+% read input files --
+infile = {infile1, infile2};
+infos = cell(2,1);
+data = cell(2,1);
+for iNum = 1:2
+    [sts_load_data, infos, data] = pspm_load_data(infile{iNum});
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_merge.m:56]: Thanks for spotting this; `pspm_cfg_run_merge` is fixed now.

File: src/pspm_cfg/pspm_cfg_run_rename.m
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_cfg/pspm_cfg_run_rename.m:1]: Thank you, this is fixed now.

**Review by teddychao — APPROVED** [review]

File: src/pspm_combine_markerchannels.m [hunk 12]
```diff
@@ -1,26 +1,29 @@
-function sts = pspm_combine_markerchannels(datafile, options)
+function [sts, outchannel] = pspm_combine_markerchannels(datafile, options)
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_combine_markerchannels.m:1]: GUI dependencies & tests will be updated in future PR, no change required for now.

File: src/pspm_convert_ecg2hb.m [hunk 13]
```diff
@@ -1,10 +1,10 @@
-function [sts,infos] = pspm_convert_ecg2hb(fn, options)
+function [sts,outchannel,debug_info] = pspm_convert_ecg2hb(fn, options)
 % ● Description
 %   pspm_convert_ecg2hb identifies the position of QRS complexes in ECG data and
 %   writes them as heart beat channel into the datafile. This function
 %   implements the algorithm by Pan & Tompkins (1985) with some adjustments.
 % ● Format
-%   sts = pspm_convert_ecg2hb(fn, options)
+%   [sts, channel_index, quality_info] = pspm_convert_ecg2hb(fn, options)
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_convert_ecg2hb.m:7]: Stylistic issue, no change required.

File: src/pspm_cfg/pspm_cfg_run_pp_emg_data.m [hunk 14]
```diff
@@ -7,9 +7,9 @@
 options.mains_freq = job.options(1).mains_freq;
 options.channel_action = job.options(1).chan_action;
 options.channel = pspm_cfg_channel_selector('run', job.options(1).chan);
-[sts, output] = pspm_emg_pp(job.datafile{1}, options);
+[sts, outchannel] = pspm_emg_pp(job.datafile{1}, options);
 if sts == 1
-  out = {output.channel};
+  out = {outchannel};
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply src/pspm_cfg/pspm_cfg_run_pp_emg_data.m:12]: GUI dependencies will be updated in a future PR.

File: test/pspm_convert_gaze_test.m [hunk 15]
```diff
@@ -31,8 +31,8 @@ function backup(this)
       import{end + 1}.type = 'gaze_y_l';
       import{end + 1}.type = 'marker';
       options = struct();
-      this.fn = pspm_import(this.raw_input_filename, 'eyelink', import, options);
-      this.fn = this.fn{1};
+      [sts, this.fn] = pspm_import(this.raw_input_filename, 'eyelink', import, options);
+      this.fn = this.fn;
```
[the remark this answers is not in the document]
Comment (dominikbach) [reply test/pspm_convert_gaze_test.m:35]: Fixed.

File: src/pspm_cfg/pspm_cfg_run_pp_heart_data.m (hunk 1 above)
Comment (dominikbach) [reply src/pspm_cfg/pspm_cfg_run_pp_heart_data.m:31]: ?

**Pull request merged**